apm-agent-plugins/apm-struts-plugin/src/main/java/co/elastic/apm/agent/struts/Struts2TransactionNameInstrumentation.java

```diff
@@ -0,0 +1,59 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2021 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.struts;
+
+import co.elastic.apm.agent.sdk.ElasticApmInstrumentation;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import java.util.Collection;
+
+import static java.util.Collections.singletonList;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+public class Struts2TransactionNameInstrumentation extends ElasticApmInstrumentation {
+
+    @Override
+    public final ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("com.opensymphony.xwork2.ActionProxy"));
```

Comment:
`hasSuperType` is quite expensive matching, please override `getTypeMatcherPreFilter` to limit the set of classes that will be searched. If you look for other usages of this type matcher they almost always use the pre-filter.

Reply:
The default implementation is located in org.apache.struts2.factory, so I used that package as a filter. The downside ist that instrumenting custom implementations of the ActionProxy class (outside that package) doesn't work any longer. But as I don't think that feature is used that much, that is probably ok.

Reply:
In that case, using a similar strategy like we have in `co.elastic.apm.agent.pluginapi.CaptureSpanInstrumentation#getTypeMatcher`, so you can include the application classes and the already known implementations, but instead do that in the `getTypeMatcherPreFilter` method.

```
    @Override
    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
        return isInAnyPackage(config.getApplicationPackages(), ElementMatchers.<NamedElement>none())
            .and(not(isProxy()))
            .and(declaresMethod(getMethodMatcher()));
    }
```

Reply:
so:
```
    @Override
    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
        return named("org.apache.struts2.factory.StrutsActionProxy")
            .or(isInAnyPackage(config.getApplicationPackages(), none()));
    }

    @Override
    public final ElementMatcher<? super TypeDescription> getTypeMatcher() {
        return hasSuperType(named("com.opensymphony.xwork2.ActionProxy"));
    }
```
?

Reply:
@SylvainJuge is the above ok?

Reply:
Sorry, I gave you to the wrong example here:

Doing something similar to 
`co.elastic.apm.agent.jaxrs.JaxRsTransactionNameInstrumentation#getTypeMatcherPreFilter` is a better strategy as it will work (albeit slow) if `application_packages` is not set.

Thus in your case it could be like this:

```
    @Override
    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
        return named("org.apache.struts2.factory.StrutsActionProxy")
            .or(isInAnyPackage(config.getApplicationPackages(), any()));
    }

    @Override
    public final ElementMatcher<? super TypeDescription> getTypeMatcher() {
        return hasSuperType(named("com.opensymphony.xwork2.ActionProxy"));
    }
```